File: backend/src/application/agent/OptimizationProviderRegistry.ts
import { Environment } from "../../config/Environment.js";
import { LlmProviderConfigRepository } from "../../domain/repositories/LlmProviderConfigRepository.js";
import { FallbackOptimizationProvider } from "./FallbackOptimizationProvider.js";
import { LlmOptimizationProvider } from "./LlmOptimizationProvider.js";
import { OptimizationProvider } from "./OptimizationProvider.js";

export class OptimizationProviderRegistry {
  constructor(
    private readonly environment: Environment,
    private readonly providerConfigs: LlmProviderConfigRepository
  ) {}

  async providers(): Promise<OptimizationProvider[]> {
    const fallback = new FallbackOptimizationProvider();
    const config = await this.providerConfigs.findDefaultEnabled();

    if (config?.provider === "openai-compatible") {
      return [
        new LlmOptimizationProvider(this.environment, config),
        fallback
      ];
    }

    return [fallback];
  }
}

File: app/services/agent/optimization_provider_registry.rb
module Agent
  class OptimizationProviderRegistry
    def providers
      config = LlmProviderConfig.default_enabled
      fallback = FallbackOptimizationProvider.new

      if config&.provider == "openai-compatible"
        [LlmOptimizationProvider.new(config), fallback]
      else
        [fallback]
      end
    end
  end
end
